packages/story-editor/src/components/panels/document/taxonomies/taxonomies.js

```diff
@@ -18,18 +18,124 @@
  * External dependencies
  */
 import { __ } from '@web-stories-wp/i18n';
+import {
+  useCallback,
+  useDebouncedCallback,
+  useMemo,
+} from '@web-stories-wp/react';
 /**
  * Internal dependencies
  */
+import { useTaxonomy } from '../../../../app/taxonomy';
+import cleanForSlug from '../../../../utils/cleanForSlug';
+import Tags from '../../../form/tags';
 import { SimplePanel } from '../../panel';
 
-function TaxonomiesPanel({ ...props }) {
+function TaxonomiesPanel(props) {
+  const {
+    taxonomies,
+    createTerm,
+    addSearchResultsToCache,
+    freeformCache,
+    setSelectedFreeformSlugs,
+    selectedFreeformSlugs,
+  } = useTaxonomy(
+    ({
+      state: { taxonomies, freeformCache, selectedFreeformSlugs },
+      actions: {
+        createTerm,
+        addSearchResultsToCache,
+        setSelectedFreeformSlugs,
+      },
+    }) => ({
+      taxonomies,
+      freeformCache,
+      createTerm,
+      addSearchResultsToCache,
+      setSelectedFreeformSlugs,
+      selectedFreeformSlugs,
+    })
+  );
+
+  const _handleFreeformTermsChange = useCallback(
+    (taxonomy) => (termNames) => {
+      termNames.forEach((termName) => createTerm(taxonomy, termName));
+      setSelectedFreeformSlugs(
+        taxonomy,
+        termNames.map((termName) => cleanForSlug(termName))
+      );
+    },
+    [createTerm, setSelectedFreeformSlugs]
+  );
+
+  const _handleFreeformInputChange = useDebouncedCallback(
+    (taxonomy) => (value) => {
+      if (value.length < 3) {
+        return;
+      }
+      addSearchResultsToCache(taxonomy, value);
+    },
+    1000
+  );
+
+  const _termDisplayTransformer = useCallback(
+    (taxonomy) => (tagName) =>
+      freeformCache[taxonomy]?.[cleanForSlug(tagName)]?.name,
+    [freeformCache]
+  );
+
+  // We want to prevent curried functions from creating
+  // a new function on every render so we build them with
+  // memoized args here instead of in the render
+  const taxonomyHandlerTuples = useMemo(
+    () =>
+      (taxonomies || []).map((taxonomy) => [
+        taxonomy,
+        // handlers
+        {
+          handleFreeformTermsChange: _handleFreeformTermsChange(taxonomy),
+          handleFreeformInputChange: _handleFreeformInputChange(taxonomy),
+          termDisplayTransformer: _termDisplayTransformer(taxonomy),
+        },
+      ]),
+    [
+      taxonomies,
+      _handleFreeformTermsChange,
+      _handleFreeformInputChange,
+      _termDisplayTransformer,
+    ]
+  );
+
   return (
     <SimplePanel
       name="taxonomies"
       title={__('Categories and Tags', 'web-stories')}
       {...props}
-    />
+    >
+      {taxonomyHandlerTuples.map(([taxonomy, handlers]) =>
+        // TODO support all taxonomies and differentiate
+        // input component based on `taxonomy.hierarchical`
+        taxonomy.slug === 'story-tag' ? (
+          <div key={taxonomy.slug}>
+            <Tags.Label htmlFor={`${taxonomy.slug}-input`}>
+              {__('Add New Term', 'web-stories')}
```

Comment:
Does this need to respect the labels register the taxonomy labels. See https://developer.wordpress.org/reference/functions/get_taxonomy_labels/#return

Reply:
These labels could be added to the taxononmies endpoint if that is at all helpful.

Reply:
Oh yes. I was looking for something like this and just had this in here temporarily. Thanks!

Reply:
yea adding them to the endpoint would be awesome. Will be easy to update here if it's already in the taxonomy response

Reply:
The labels information is already available from the taxonomies endpoint. No need to reinvent the wheel. You just need to pass `&context=edit` in the `getTaxonomies` API request to get all the taxonomies including the labels.

With edit context, the user needs to have the `assign_terms` capability for a taxonomy to see it & its labels. This relates a bit to #9011